### PR TITLE
[sdk/dotnet] Add PluginDownloadURL

### LINKED ISSUE #8689: [sdk/dotnet] Support for custom server urls
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details
Provide the .NET SDK hookup for #8544.

### PR DESCRIPTION
**Pull request by iwahbe**
# Description

Fixes #8689 

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

File: pkg/codegen/dotnet/templates.go [hunk 1]
```diff
@@ -85,6 +86,17 @@ namespace {{.Namespace}}
             };
         }
 
+        public static InvokeOptions WithDefaults(this InvokeOptions? options)
+        {
+            return new InvokeOptions
+            {
+                Parent = options?.Parent,
+                Provider = options?.Provider,
+                Version = options?.Version != null ? options?.Version : Version,
```
Comment (justinvp) [inline pkg/codegen/dotnet/templates.go:95]: Nit: Could use the [null-coalescing operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-coalescing-operator) `??`:

```suggestion
                Version = options?.Version ?? Version,
```

File: sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs [hunk 2]
```diff
@@ -112,6 +112,7 @@ private async Task<SerializationResult> InvokeRawAsync(string token, Serializati
                 Version = options?.Version ?? "",
                 Args = serialized,
                 AcceptResources = !_disableResourceReferences,
+                PluginDownloadURL = options?.PluginDownloadURL ?? "",
```
Comment (justinvp) [inline sdk/dotnet/Pulumi/Deployment/Deployment_Invoke.cs:115]: Need to pass the URL set on the `Resource`'s `_pluginDownloadURL` field to `CallRequest` in `Deployment_Call.cs` as well:

https://github.com/pulumi/pulumi/blob/3715afa5907db5d80dab8c107a8c03f0268169c8/sdk/dotnet/Pulumi/Deployment/Deployment_Call.cs#L79

File: pkg/codegen/dotnet/gen.go [hunk 3]
```diff
@@ -1158,10 +1158,23 @@ func (mod *modContext) genResource(w io.Writer, r *schema.Resource) error {
 		if fun.DeprecationMessage != "" {
 			fmt.Fprintf(w, "        [Obsolete(@\"%s\")]\n", strings.ReplaceAll(fun.DeprecationMessage, `"`, `""`))
 		}
-
 		fmt.Fprintf(w, "        public %s %s(%s)\n", returnType, methodName, argsParamDef)
-		fmt.Fprintf(w, "            => Pulumi.Deployment.Instance.Call%s(\"%s\", %s, this)%s;\n",
-			typeParameter, fun.Token, argsParamRef, lift)
+		fmt.Fprintf(w, "        {\n")
+		fmt.Fprintf(w, "            var defaultOptions = new CallOptions\n")
```
Comment (justinvp) [inline pkg/codegen/dotnet/gen.go:1163]: There shouldn't be a need for these codegen changes.

The implementation of `Pulumi.Deployment.Instance.Call` should be updated to get the download URL saved on the `Resource` instance:

https://github.com/pulumi/pulumi/blob/12ec51a5f3d19971dbbf046516ab40fa290b0346/sdk/dotnet/Pulumi/Deployment/Deployment_Call.cs#L64-L70

Comment (iwahbe) [reply pkg/codegen/dotnet/gen.go]: I think you are right that the codegen change is unnecessary. I'll revert it.

**Review by justinvp — APPROVED** [review]
LGTM

**Comment by iwahbe** [comment]
Failing test is windows (OOM). This is a known issue, so I'm merging anyway.

**Pull request merged**